utils/s3/client.cc

```diff
@@ -338,6 +351,105 @@ data_sink client::make_upload_sink(sstring object_name) {
     return data_sink(std::make_unique<upload_sink>(shared_from_this(), std::move(object_name)));
 }
 
+class client::readable_file : public file_impl {
```

Comment:
It's not in fact read-_only_. It's created by sstable code with flags::rw with no errors and pretends to be writable too. The only reason why it doesn't throw after it is because sstables code doesn't write directly into it, but instead creates output_stream for writing and all those places are patched to create stream via storage virtual call. I tried to avoid these hacks with scylladb/seastar#1419